### ISSUE TITLE
Excessive stack frame usage in dmu_recv_stream in debug mode 

### ISSUE THREAD

**Issue by lorddoskias** [zfsonlinux/zfs#4565]
When compiling trunk zfs with `--enable-debug ` I get the following warning: 

```
/home/nborisov/projects/zfs/myfork/zfs/module/zfs/dmu_send.c: In function ‘dmu_recv_stream’:
/home/nborisov/projects/zfs/myfork/zfs/module/zfs/dmu_send.c:2502:1: error: the frame size of 1256 bytes is larger than 1024 bytes [-Werror=frame-larger-than=]
```

My gcc version is `gcc (Ubuntu 5.3.1-14ubuntu2) 5.3.1 20160413` I also tried with `gcc (Ubuntu/Linaro 4.7.4-3ubuntu12) 4.7.4` and same error happens.

**Comment by behlendorf** [comment]
Yup, this function is definitely getting stack heavy.  We'll want to move `struct receive_arg` and `struct receive_writer_arg` on to the heap.

**Comment by lorddoskias** [comment]
Fixed in [04bc461062df8e272aa11268668140](https://github.com/zfsonlinux/zfs/commit/04bc461062df8e272aa11268668140c2976ac823)
